### ISSUE TITLE
generic article: nf-core/nextflow offline deployment

### ISSUE THREAD

**Issue by ptrebert** [core-unit-bioinformatics/knowledge-base#2]
as discussed, we need a generic article on offline deployment of nextflow (or rather: nf-core?) workflows, if that is the same process for enough workflows.

**Comment by svenbioinf** [comment]
Should the wiki also cover below points or is this better suited for an internal HPC discussion:
nf-core pipelines need...
- common folder for singularity images
- common folder for references
- common folder structure (run folder, config file location..., results)
- location for cubi metadata (consistent for snakemake and nextflow)
- nf-core on offline hpc wiki (explain how to let nf-core run without inet connection)
- how to use automatic sample sheet script (if we have one)

**Comment by ptrebert** [comment]
I think the final answer will have to be developed when the first draft(s) of the article can be discussed. Right now, I think everything that would need to be configured on any *other* offline system as well should be covered. The concrete pointers for our local infrastructure should then live in a separate repo in the githhu (linked in the article) similar to the cubi fork of the snakemake-utils

**Comment by svenbioinf** [comment]
I started a readme for nf-core/rnaseq which can be used as basis but needs to be more generalized:
[here](https://github.com/core-unit-bioinformatics/workflow-nxf-rnaseq/tree/master/cubi)

**Comment by ptrebert** [comment]
1) Maybe I am missing something too obvious, but I just created a Conda environment with `nf-core 2.7.2`, and it seems that `nextflow` is not installed as a dependency?! If that is indeed the case, this section should start by defining how to create an execution-ready environment:

https://github.com/core-unit-bioinformatics/knowledge-base/blob/add-nextflow-config-article/setup-nfcore-workflow.md#running-nf-core-workflows-

2) I do not understand why the resource caps are set this way. This is well below the actual maximum on HILBERT?!

3) the point `2.4 Optional: Modify resources for specific processes` confuses me because I would naively assume that Nextflow can automatically scale resources up when restarting a job? If that is the case, this should be mentioned here as it should be the preferred way.

**Comment by svenwillger** [comment]
1. https://github.com/core-unit-bioinformatics/knowledge-base/blob/694dd0b6bb76ddbd227bdc2938029fa5dd06adba/setup-nfcore-workflow.md?plain=1#L16-L20 
Just to second Peter's last comment, in this section it should be mentioned that the best way to run Nextflow is to create a conda environment because on HILBERT or any HPC that is the only way it works and even on a local machine installing it via pip, which is mentioned, it can cause annoying dependency issues for several packages, so I would suggest to not include this option in the description.

A list of the required packages for this conda env should be provided, e.g.:
- mamba=1.5.6 (strongly recommended)
- nextflow=23.10.1
- nf-core=2.11.1
- singularity=2.4.6

Also add a brief description how to set up the `NXF_SINGULARITY_CACHEDIR` (you can still link to the detailed documentation on the nf-core webpage, although the webpage is outdated regarding the `--container` option, see below)

2. https://github.com/core-unit-bioinformatics/knowledge-base/blob/694dd0b6bb76ddbd227bdc2938029fa5dd06adba/setup-nfcore-workflow.md?plain=1#L37

- I would use a different workflow other than sarek because it's huge and requires a lot of packages and external databases. Use a small workflow e.g. `bamtofastq` (https://nf-co.re/bamtofastq/2.1.0)
- The nf-core options have changed since you wrote this markdown and the option `--container (-c)` doesn't exist anymore. This makes it more important that you mention in the first step which version(s) you are using. The new version would have to look like this:
`nf-core download bamtofastq --revision 2.1.0 --compress none --container-system singularity`
- Since we have to download the workflow locally and then copy it to an HPC this section should point out that you want to `copy` the images instead of the default setting of `amend` them

3. https://github.com/core-unit-bioinformatics/knowledge-base/blob/694dd0b6bb76ddbd227bdc2938029fa5dd06adba/setup-nfcore-workflow.md?plain=1#L53
This becomes less important if we are not using sarek but if we want to keep a description how to download external reference datasets you have to fix the command because it's missing a space to separate the source from the destination:
`aws s3 --no-sign-request --region eu-west-1 sync s3://ngi-igenomes/igenomes/Homo_sapiens/GATK/GRCh38/  db/references/Homo_sapiens/GATK/GRCh38/`

4. https://github.com/core-unit-bioinformatics/knowledge-base/blob/694dd0b6bb76ddbd227bdc2938029fa5dd06adba/setup-nfcore-workflow.md?plain=1#L62
The link to the `nextflow-contig.md` needs to be updated.

**Comment by sci-kai** [comment]
Thanks for your comments @svenwillger and @ptrebert 

I added the following changes in commit 24bb272 :

1. Add section about installation of nextflow and nf-core in conda environment. Also adding a yml file for reproduction.
2. Added a note about the singularity cache dir specification
3. Changed the example to bamtofastq (to speed up download processes)
4. Adjusting the resource section. @ptrebert was right, I first recommend to increase the maximum caps and added the possibility to modify the resources to avoid long runtimes and speed up the workflow. I do not know the maximum resources of HILBERT and haven't changed that.
5. Fixed minor errors
